### PR TITLE
Add ConfigMap/Secret revision labels for environment variables

### LINKED ISSUE #13670: machine-controller is not restarted when cloud credentials are updated
### What happened?

When updating the cloud credentials in the `Cluster` resource, they're rolled out to all the correct places (cloud-config, etc.) Furthermore things like the CCM and I think the CSI are restarted. However, the `machine-controller` and `machine-controller-webhook` are not. Thus, if the old credentials don't work anymore, machines can't be added or removed anymore.

### Expected behavior

Changing cloud credentials should restart the `machine-controller` pod.

### How to reproduce the issue?

- change cloud credentials
- observe `machine-controller` and `machine-controller-webhook` pods not restarted
- delete old credentials from cloud
- observe machines can't be added or deleted anymore
- restart `machine-controller` and `machine-controller-webhook`
- oberseve machines can be deleted and added again

### How is your environment configured?

- KKP version: 2.25.8
- Shared or separate master/seed clusters?: shared

### Provide your KKP manifest here (if applicable)

N/A

</details>

### What cloud provider are you running on?

OpenStack

### What operating system are you running in your user cluster?

Ubuntu 22.04

### PR DESCRIPTION
**Pull request by xrstf**
**What this PR does / why we need it**:
KKP adds labels to the PodTemplates of Deployments, one for each _mounted_ ConfigMap/Secret. When the ConfigMap/Secret changes, after a reconciliation this will force Kubernetes to rotate the Deployments.

However KKP did not consider ConfigMap or Secret references for environment variables in containers. So for example the cloud-credentials are never mounted, but only referred to via environment variables and so changing the credentials will not lead to a rotation.

This PR begun by extending the existing VolumeRevisionLabels functionality. But this function is called manually in each individual reconciler, and with the new code that would mean each call has to also hand over the containers and init containers. This seemed tedious and so I decided to instead turn this explicit function call into a re-usable ObjectModifier.

This new ObjectModifier lives in `pkg/resources/reconciling/modifier`, together with the other two commonly used modifiers from the KKP operator: VersionLabel and Ownership.

All reconcilers that used the old explicit function call were then rewritten to simply not do it anymore, and instead the controllers will initialize the necessary modifiers and apply them transparently. This makes each reconciler a bit smaller and removes the akward need to initialize the volumes very early only because we needed them to setup the object labels.

However the old function added two additional labels: `app` (always identical to the Deployment's name) and `cluster` (KKP Cluster name). adding these labels is now also done with a modifier, the new ControlplaneComponent modifier.

As a neat side effect, the volume revision labels (now called related revisions labels) can be applied to StatefulSets and even DaemonSets.

**NB:** The code to determine the revision labels assumes, and always has, that the referred to objects _exist already_. This means we cannot reconcile a Deployment that mounts a non-existing ConfigMap or uses it for an environment variable (`optional` is respected, and a missing optional reference will simply omit the label).

This goes a bit against Kubernetes' eventual consistency system, but has never been an actual issue in KKP. KKP creates all related objects itself, and does so always in-order. For example for certificates we do not rely on cert-manager to _eventually_ create a Secret, but instead we create the certs in Go ourselves. Additionally the reconciling framework ensures newly created objects are in the cache before proceeding, so once ReconcileSecrets() has finished, we can be certain the Secrets all exist.

So I think we do not need to change the behaviour here. From first testing and looking through the code, we always do things in the "proper" order.

**Which issue(s) this PR fixes**:
Fixes #13670, one step closer towards #13563

**What type of PR is this?**
/kind feature

**Does this PR introduce a user-facing change? Then add your Release Note here**:
```release-note
Control plane components will now automatically be rotated when usercluster cloud credentials are changed.
```

**Documentation**:
```documentation
NONE
```

### CODE REVIEW

File: pkg/resources/reconciling/modifier/related_revisions.go [hunk 1]
```diff
@@ -0,0 +1,217 @@
...
+			return nil, fmt.Errorf("failed to determine revision labels: %w", err)
+		}
+
+	case *appsv1.DaemonSet:
+		if err := addRevisionLabelsToPodSpec(ctx, client, asserted.Namespace, &asserted.Spec.Template); err != nil {
+			return nil, fmt.Errorf("failed to determine revision labels: %w", err)
+		}
+	}
+
+	return obj, nil
```
Comment (xmudrii) [inline pkg/resources/reconciling/modifier/related_revisions.go:69]: Why we don't have a default case that's going to throw an error if the object is not one of these three types? We have something like that in the version label modifier. My worry here is that someone might for any reason try to use this modifier on some other object type, and that would do nothing, but they might not expect that.

Comment (xrstf) [reply pkg/resources/reconciling/modifier/related_revisions.go:69]: Agreed, I added a panic().

**Review by xmudrii — APPROVED** [review]
/lgtm
/approve

**Pull request merged**